### ISSUE TITLE
`manual_let_else` should reuse the identifier from the original let statement

### ISSUE THREAD

**Issue by Serial-ATA** [rust-lang/rust-clippy#9939]
### Summary

The lint will emit an invalid suggestion when the binding name in the pattern differs from the name used in the let statement.

### Reproducer

I tried this code:

```rust
#![warn(clippy::manual_let_else)]

fn main() {
    let foo = Some(1);

    let value = match foo {
        Some(v) => v,
        _ => return,
    };
    
    println!("We have a value of: {value}");
}
```

I expected to see this happen:

```rust
#![warn(clippy::manual_let_else)]

fn main() {
    let foo = Some(1);

    let Some(value) = foo else { return };
    
    println!("We have a value of: {value}");
}
```
This is the suggestion it should emit, reusing the name `value`.

Instead, this happened:

```rust
#![warn(clippy::manual_let_else)]

fn main() {
    let foo = Some(1);

	// It reuses the name `v`, which is not expected later on
    let Some(v) = foo else { return };
    
    println!("We have a value of: {value}");
}
```
This suggestion causes an error:
```
error[[E0425]](https://doc.rust-lang.org/nightly/error-index.html#E0425): cannot find value `value` in this scope
 --> src/main.rs:8:36
  |
8 |     println!("We have a value of: {value}");
  |                                    ^^^^^ not found in this scope
```

### Version

```text
rustc 1.67.0-nightly (70f8737b2 2022-11-23)
binary: rustc
commit-hash: 70f8737b2f5d3bf7d6b784fad00b663b7ff9feda
commit-date: 2022-11-23
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by andersk** [comment]
Just to note a corner case where *both* identifiers will need to be present:

```rust
enum Foo {
    Yes { v: i32 },
    No,
}
let foo = Foo::Yes { v: 1 };
let value = match foo {
    Foo::Yes { v } => v,
    _ => return,
};
println!("We have a value of: {value}");
```

The current incorrect suggestion

```rust
let Foo::Yes { v } = foo else { return };
```

will need to become

```rust
let Foo::Yes { v: value } = foo else { return };
```

**Comment by est31** [comment]
There is also the shadowing corner case, e.g.:

```Rust
enum Foo {
    Yes { foo: u32, bar: u32 },
    No,
}
let _foo = 42;
let bar = if let Foo::Yes { foo: _foo, bar } = make_foo() {
    bar
} else {
    return
};
println!("{_foo}");
```

You can't change that to:

```Rust
let Foo::Yes { foo: _foo, bar } = make_foo() else { return };
```

Because that might print a different `_foo` value :).

The issue exists however only in code that is either receiving one of the `unused` rustc lints, or code where you use a variable with a leading `_`, which is a bit weird on its own. It's thus a corner case, but there is the possibility to change behaviour!

**Comment by smoelius** [comment]
A related example, if it helps:
```rust
#![warn(clippy::pedantic)]
#![allow(clippy::unnecessary_wraps)]

use std::io::Result;

fn foo() -> Result<Option<(u32, u32)>> {
    Ok(Some((1, 2)))
}

fn main() -> Result<()> {
    let (x, y) = if let Some(pair) = foo()? {
        pair
    } else {
        return Ok(());
    };
    println!("{x}, {y}");
    Ok(())
}
```
The suggestion removes the bindings for `x` and `y`:
```rust
     let Some(pair) = foo()? else {
         return Ok(());
     };
     println!("{x}, {y}"); // `x`, `y` not found in this scope
```
(Aside: I genuinely _love_ this lint! ❤️)
